xarray/core/dataset.py

```diff
@@ -1933,6 +1933,47 @@ def from_dataframe(cls, dataframe):
             obj[name] = (dims, data)
         return obj
 
+    def to_dict(self):
+        """Convert this dataset to a dictionary following xarray naming conventions.
+        """
+        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': list(self.dims), 
+             'data_vars': {}}
+
+        for k in self.coords:
+            d['coords'].update({k: {'data': self[k].values.tolist(),
+                                    'dims': list(self[k].dims),
+                                    'attrs': dict(self[k].attrs)}})
+        for k in self.data_vars:
+            d['data_vars'].update({k: {'data': self[k].values.tolist(),
+                                       'dims': list(self[k].dims),
+                                       'attrs': dict(self[k].attrs)}})
+        return d
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Convert a dictionary into an xarray.Dataset.
+        """
+        obj = cls()
+
+        dims=OrderedDict([(k, d['coords'][k]) for k in d['dims']])
+        for dim, dim_d in dims.items():
+            obj[dim] = (dim_d['dims'], dim_d['data'], dim_d['attrs'])
+
+        for var, var_d in d['data_vars'].items():
+            obj[var] = (var_d['dims'], var_d['data'], var_d['attrs'])
+
+        # what it coords aren't dims?
```

Comment:
A slightly cleaner way to solve this is to loop over `d['coords'].items()` instead of looping of `dims.items()` separately.